# Import and Read the Data

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime

import warnings
warnings.simplefilter(action='ignore', category= FutureWarning)

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
schedule_df = pd.read_csv("/content/drive/MyDrive/Braincore/KRL/schedule.csv", delimiter=",")
station_df = pd.read_csv("/content/drive/MyDrive/Braincore/KRL/station.csv", delimiter=",")

In [ ]:
awalloglat= pd.read_csv("/content/drive/MyDrive/Braincore/KRL/awallanglot - ANCOL', 'ANGKE', 'BRAMBANAN', 'BANDUNG', 'BOJONGG....csv")
akhirloglat= pd.read_csv("/content/drive/MyDrive/Braincore/KRL/akhirloglat.csv")

In [ ]:
df = pd.read_csv("/content/drive/MyDrive/Braincore/KRL/mergelonglat.csv")

In [ ]:
time= pd.read_csv("/content/drive/MyDrive/Braincore/KRL/CleanKRLData - CleanKRLData.csv")

https://colab.research.google.com/drive/1D5CoW1a4JPvabMH55XMeXGuE3mD4fTtJ?usp=sharing

In [ ]:
data = pd.merge(schedule_df, station_df, left_on="station_id", right_on="id", how="inner")
data.drop(['id_y','updated_at_x','updated_at_y'], axis=1, inplace=True)
data['origin'] = data['route'].apply(lambda x: x.split("-")[0])
data = data.rename(columns={"name": "name"})

In [ ]:
schedule_df.head(3)

id station_id train_id                        line  \
0  AC-2400         AC     2400  COMMUTER LINE TANJUNGPRIUK   
1  AC-2401         AC     2401  COMMUTER LINE TANJUNGPRIUK   
2  AC-2402         AC     2402  COMMUTER LINE TANJUNGPRIUK   

                      route    color   destination time_estimated  \
0  JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK       06:07:00   
1  TANJUNGPRIUK-JAKARTAKOTA  #DD0067   JAKARTAKOTA       06:34:00   
2  JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK       06:24:00   

  destination_time                updated_at  
0         06:16:00  2024-03-16T17:00:08.063Z  
1         06:42:00  2024-03-16T17:00:08.063Z  
2         06:33:00  2024-03-16T17:00:08.063Z

In [ ]:
# df['origin'].head(3)

In [ ]:
station_df.head(3)

id       name  daop  fg_enable  have_schedule                updated_at
0   AC      ANCOL     1          1           True  2024-03-10T15:15:19.997Z
1   AK      ANGKE     1          1           True  2024-03-10T15:15:19.997Z
2  BBN  BRAMBANAN     6          1           True  2024-03-10T15:15:19.997Z

In [ ]:
schedule_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18065 entries, 0 to 18064
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   id                18065 non-null  object
 1   station_id        18065 non-null  object
 2   train_id          18065 non-null  object
 3   line              18065 non-null  object
 4   route             18065 non-null  object
 5   color             18065 non-null  object
 6   destination       18065 non-null  object
 7   time_estimated    18065 non-null  object
 8   destination_time  18065 non-null  object
 9   updated_at        18065 non-null  object
dtypes: object(10)
memory usage: 1.4+ MB


In [ ]:
station_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 138 entries, 0 to 137
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   id             138 non-null    object
 1   name           138 non-null    object
 2   daop           138 non-null    int64 
 3   fg_enable      138 non-null    int64 
 4   have_schedule  138 non-null    bool  
 5   updated_at     138 non-null    object
dtypes: bool(1), int64(2), object(3)
memory usage: 5.7+ KB


In [ ]:
awalloglat.head(3)

name AwalLatitude      AwalLongi
0      ANCOL  -61.257.714  1.068.032.377
1      ANGKE       -6.135    106.811.111
2  BRAMBANAN   -7.782.344    110.361.298

In [ ]:
print("Unique destinations in akhirloglat:")
print(awalloglat['name'].unique())
print("\nUnique destinations in schedule:")
print(station_df['name'].unique())

Unique destinations in akhirloglat:
['ANCOL' 'ANGKE' 'BRAMBANAN' 'BANDUNG' 'BOJONGGEDE' 'BEKASI' 'BEKASITIMUR'
 'BOJONGINDAH' 'BOGOR' 'BATUCEPER' 'BANDARASOEKARNOHATTA' 'BUARAN'
 'CIGANEA' 'CIBATU' 'CIBINONG' 'CIBUNGUR' 'CICAYUR' 'CICALENGKA'
 'CIKADONGDONG' 'CEPER' 'CIBITUNG' 'CILEJIT' 'CIKINI' 'CIKAMPEK'
 'CIKARANG' 'CIKOYA' 'CILAME' 'CILEBUT' 'CIMINDI' 'CIMAHI' 'CIMEKAR'
 'CISAUK' 'CITAYAM' 'CIKUDAPATEUH' 'CITERAS' 'CAWANG' 'DARU' 'DELANGGU'
 'DEPOK' 'DEPOKBARU' 'DURENKALIBATA' 'DURI' 'DAWUAN' 'GEDEBAGE'
 'GONDANGDIA' 'GROGOL' 'GADOBANGKONG' 'GARUT' 'GANGSENTIONG' 'GAWOK'
 'HAURPUGUR' 'JAKARTAKOTA' 'JAYAKARTA' 'JURANGMANGU' 'JENAR' 'JATINEGARA'
 'JUANDA' 'KIARACONDONG' 'KARET' 'KEBAYORAN' 'KEDUNGGEDEH' 'KALIDERES'
 'KLENDER' 'KLENDERBARU' 'KLARI' 'KEMAYORAN' 'KRAMAT' 'KOSAMBI'
 'KAMPUNGBANDAN' 'KARANGSARI' 'KRANJI' 'KLATEN' 'KUTOARJO' 'KARAWANG'
 'LEBAKJERO' 'LELES' 'LEMAHABANG' 'LENTENGAGUNG' 'LEUWIGOONG'
 'LEMPUYANGAN' 'MANGGABESAR' 'MAGUWO' 'MAJA' 'MANGGARAI' 'MASWATI'
 'MATRAMAN

In [ ]:
awalloglat.shape

(137, 3)

In [ ]:
akhirloglat.head(3)

destination  AkhirLatitude  AkhirLongitude
0  TANJUNGPRIUK      -61280924      1068270132
1   JAKARTAKOTA      -61257714      1068032377
2        TAMBUN       -6175694       106924778

In [ ]:
# print("Unique destinations in akhirloglat:")
# print(akhirloglat['destination'].unique())
# print("\nUnique destinations in schedule:")
# print(schedule_df['destination'].unique())

In [ ]:
# # merging the schedule and station dataset
# df = pd.merge(schedule_df, station_df, left_on="station_id", right_on="id", how="inner")
# df.drop("id_y", axis=1, inplace=True)

In [ ]:
# # Merge awalloglat with station based on the 'name' column
# merged_awal = pd.merge(awalloglat, station_df, left_on="name", right_on="name", how="inner")
# merged_awal.head(3)

In [ ]:
# # 137 stasiun
# merged_awal.shape

In [ ]:
# # Merge akhirloglat with station based on the 'destination' column
# merged_akhir = pd.merge(akhirloglat, station_df, left_on="destination", right_on="name", how="inner")
# merged_akhir.head(3)

In [ ]:
# # 23 Destinasi
# merged_akhir.shape

Diatas memberi longitute latitude berdasarkan nama statiun keberngkatan dengan tujuan, tanpa mengetahui rutenya. Pada dataframe `station_df`




name	AwalLatitude	AwalLongi

destination	AkhirLatitude	AkhirLongitude

In [ ]:
# # Merge the first result with schedule based on the name of station 'station_id' column
# merged_df1 = pd.merge(merged_awal, schedule_df, left_on="name", right_on="destination", how="inner")
# merged_df1.head(3)

In [ ]:
# merged_df1.shape

In [ ]:
# # Merge the second result with schedule based on the 'destination' column
# merged_df2 = pd.merge(merged_akhir, schedule_df, on="destination", how="inner")
# merged_df2.head(3)

In [ ]:
# # Merge the first result with schedule based on the name of station 'station_id' column
# df = pd.merge(merged_df1, merged_df2, left_on="name", right_on="destination", how="inner")
# df.head(3)

In [ ]:
pd.set_option('display.max_colwidth', 100)
# data unik di tiap kolom
listItem = []
for col in df.columns :
    listItem.append( [col, df[col].nunique(), df[col].unique()])

tabel1Desc = pd.DataFrame(columns=['Column Name', 'Number of Unique', 'Unique Sample'],
                     data=listItem)
tabel1Desc

Column Name  Number of Unique  \
0         station_id                99   
1           train_id              1276   
2               line                 9   
3              route                67   
4              color                 6   
5      destination_x                23   
6     time_estimated              1261   
7   destination_time               826   
8       updated_at_x               112   
9             name_x                99   
10      AwalLatitude                60   
11         AwalLongi                48   
12            daop_x                 2   
13       fg_enable_x                 1   
14   have_schedule_x                 1   
15      updated_at_y                 1   
16     AkhirLatitude                16   
17    AkhirLongitude                15   
18                id                23   
19            daop_y                 2   
20       fg_enable_y                 1   
21   have_schedule_y                 1   
22        updated_at                 1   

                                                                                          Unique Sample  
0   [AC, JAKK, KPB, BJD, BKS, BKST, BOO, BUA, CBN, CIT, CKI, CKR, CLT, CTA, CW, DP, DPB, DRN, GDD, G...  
1   [2400, 2402, 2404, 2406, 2408, 2410, 2412, 2414, 2416, 2418, 2420, 2422, 2424, 2426, 2428, 2430,...  
2   [COMMUTER LINE TANJUNGPRIUK, COMMUTER LINE BOGOR, COMMUTER LINE CIKARANG, COMMUTER LINE TANGERAN...  
3   [JAKARTAKOTA-TANJUNGPRIUK, TANJUNGPRIUK-JAKARTAKOTA, BOGOR-JAKARTAKOTA, BOJONGGEDE-JAKARTAKOTA, ...  
4                                                [#DD0067, #E30A16, #0084D8, #623814, #F76114, #16812B]  
5   [TANJUNGPRIUK, JAKARTAKOTA, TAMBUN, CIKARANG, BEKASI, KAMPUNGBANDAN, MANGGARAI, YOGYAKARTA, PALU...  
6   [06:07:00, 06:24:00, 06:43:00, 07:01:00, 07:19:00, 07:37:00, 07:55:00, 08:13:00, 08:31:00, 08:49...  
7   [06:16:00, 06:33:00, 06:52:00, 07:10:00, 07:28:00, 07:46:00, 08:04:00, 08:22:00, 08:40:00, 08:58...  
8   [2024-03-16T17:00:08.063Z, 2024-03-16T17:00:12.377Z, 2024-03-16T17:00:14.728Z, 2024-03-16T17:00:...  
9   [ANCOL, JAKARTAKOTA, KAMPUNGBANDAN, BOJONGGEDE, BEKASI, BEKASITIMUR, BOGOR, BUARAN, CIBINONG, CI...  
10  [-6.12803, -6.13333, -6.166667, -6.40167, -6.22938, -6.24028, -6.60094, -6.18333, -6.52778, -6.2...  
11  [106.8032377, 106.816667, 106.875, 106.996458, 107.004444, 106.804522, 106.833333, 106.85, 106.9...  
12                                                                                               [1, 6]  
13                                                                                                  [1]  
14                                                                                               [True]  
15                                                                           [2024-03-10T15:15:19.997Z]  
16  [-6.1280924, -6.1257714, -6.175694, -6.227948, -6.229378, -6.140538, -6.219544, -7.782344, -6.60...  
17  [106.8270132, 106.8032377, 106.924778, 106.982722, 106.996458, 106.840366, 106.833142, 110.36129...  
18  [TPK, JAKK, TB, CKR, BKS, KPB, MRI, YK, PL, BOO, DP, AK, DU, TNG, BST, NMO, PRP, RK, THB, TGS, S...  
19                                                                                               [1, 6]  
20                                                                                                  [1]  
21                                                                                               [True]  
22                                                                           [2024-03-10T15:15:19.997Z]

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17756 entries, 0 to 17755
Data columns (total 23 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   station_id        17756 non-null  object 
 1   train_id          17756 non-null  object 
 2   line              17756 non-null  object 
 3   route             17756 non-null  object 
 4   color             17756 non-null  object 
 5   destination_x     17756 non-null  object 
 6   time_estimated    17756 non-null  object 
 7   destination_time  17756 non-null  object 
 8   updated_at_x      17756 non-null  object 
 9   name_x            17756 non-null  object 
 10  AwalLatitude      17756 non-null  float64
 11  AwalLongi         17756 non-null  float64
 12  daop_x            17756 non-null  int64  
 13  fg_enable_x       17756 non-null  int64  
 14  have_schedule_x   17756 non-null  bool   
 15  updated_at_y      17756 non-null  object 
 16  AkhirLatitude     17756 non-null  float6

# Project Goal



```
the number of insights obtained is at least 20 insights
```



# Data Understanding and Cleaning


schedule.csv =
```
id	: ID Stasiun-Kereta, contoh: AC-2400 sampai YK-7102, terdapat 18065 kereta
station_id : ID Stasiun, contoh: AC, terdapat 100 id stasiun
train_id : ID Kereta, contoh: 2400
line : Daerah Commuter Line/ Wilayah Operasi, terdapat 10. contoh: Commuter Line TANJUNGPRIUK, CIKARANG, YK, BOGOR, TANGERANG, BST, RANGKASBITUNG, PRAMBANAN EKSPRES (PRAMEKS), PRAMBANAN EKSPRES, MERAK.
route	: Rute kereta, terdapat 69 rute berbeda. contoh: JAKARTAKOTA-TANJUNGPRIUK
color	: Warna kereta, terdapat 6 warna berbeda
destination	:  tujuan, terdapat 25 daerah tujuan berbeda.
time_estimated :  Departure- waktu brangkat
destination_time :	Arrival- waktu datng
updated_at : Last time update
```

---

station.csv :

```
id : ID Stasiun, contoh: AC, terdapat 100 id stasiun
name : Nama Stasiun, terdapat 100 nama stasiun
daop : Daerah Operasional KAI, terdapat 3 jenis yaitu daop 1, 2, dan 6. 1 = Jakarta, 2= Bandung, 3 = cirebon, 4= semarang, 5= purwokerto, 6= jogja.
fg_enable : hanya bernilai 1
have_schedule	: Is the Station already make a schedule? TRUE or FALSE
updated_at : Last time update
**bold text**
```

note : KRL Commuter Line adalah sistem transportasi angkutan cepat komuter berbasis kereta rel listrik (KRL) yang dioperasikan oleh PT Kereta Commuter Indonesia (KAI Commuter), anak perusahaan dari PT Kereta Api Indonesia (Persero) (PT KAI)





In [ ]:
df.isna().sum().sum()

0

In [ ]:
df.duplicated().sum()

0

# Feature

`date`,`trip_duration_minutes`,`waktu`


In [ ]:
print(df['time_estimated'].head())
print(df['destination_time'].head())

0    06:07:00
1    06:24:00
2    06:43:00
3    07:01:00
4    07:19:00
Name: time_estimated, dtype: object
0    06:16:00
1    06:33:00
2    06:52:00
3    07:10:00
4    07:28:00
Name: destination_time, dtype: object


## Trip Duration| Waktu Perjalanan

In [ ]:
import pandas as pd

# Assuming your DataFrame is named 'df'

# Convert time_estimated and destination_time to datetime format
df['time_estimated'] = pd.to_datetime(df['time_estimated'])
df['destination_time'] = pd.to_datetime(df['destination_time'])

# Menghitung durasi perjalanan dari stasiun awal ke stasiun tujuan dalam menit
df['trip_duration_minutes'] = (df['destination_time'] - df['time_estimated']).dt.total_seconds() / 60
df['trip_duration_minutes'] = df['trip_duration_minutes'].astype(int)
df.head()

<ipython-input-30-42ef946ec6a7>:6: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['time_estimated'] = pd.to_datetime(df['time_estimated'])
<ipython-input-30-42ef946ec6a7>:7: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['destination_time'] = pd.to_datetime(df['destination_time'])


station_id train_id                        line                     route  \
0         AC     2400  COMMUTER LINE TANJUNGPRIUK  JAKARTAKOTA-TANJUNGPRIUK   
1         AC     2402  COMMUTER LINE TANJUNGPRIUK  JAKARTAKOTA-TANJUNGPRIUK   
2         AC     2404  COMMUTER LINE TANJUNGPRIUK  JAKARTAKOTA-TANJUNGPRIUK   
3         AC     2406  COMMUTER LINE TANJUNGPRIUK  JAKARTAKOTA-TANJUNGPRIUK   
4         AC     2408  COMMUTER LINE TANJUNGPRIUK  JAKARTAKOTA-TANJUNGPRIUK   

     color destination_x      time_estimated    destination_time  \
0  #DD0067  TANJUNGPRIUK 2024-05-08 06:07:00 2024-05-08 06:16:00   
1  #DD0067  TANJUNGPRIUK 2024-05-08 06:24:00 2024-05-08 06:33:00   
2  #DD0067  TANJUNGPRIUK 2024-05-08 06:43:00 2024-05-08 06:52:00   
3  #DD0067  TANJUNGPRIUK 2024-05-08 07:01:00 2024-05-08 07:10:00   
4  #DD0067  TANJUNGPRIUK 2024-05-08 07:19:00 2024-05-08 07:28:00   

               updated_at_x name_x  ...  have_schedule_x  \
0  2024-03-16T17:00:08.063Z  ANCOL  ...             True   
1  2024-03-16T17:00:08.063Z  ANCOL  ...             True   
2  2024-03-16T17:00:08.063Z  ANCOL  ...             True   
3  2024-03-16T17:00:08.063Z  ANCOL  ...             True   
4  2024-03-16T17:00:08.063Z  ANCOL  ...             True   

               updated_at_y  AkhirLatitude  AkhirLongitude   id daop_y  \
0  2024-03-10T15:15:19.997Z      -6.128092      106.827013  TPK      1   
1  2024-03-10T15:15:19.997Z      -6.128092      106.827013  TPK      1   
2  2024-03-10T15:15:19.997Z      -6.128092      106.827013  TPK      1   
3  2024-03-10T15:15:19.997Z      -6.128092      106.827013  TPK      1   
4  2024-03-10T15:15:19.997Z      -6.128092      106.827013  TPK      1   

   fg_enable_y  have_schedule_y                updated_at  \
0            1             True  2024-03-10T15:15:19.997Z   
1            1             True  2024-03-10T15:15:19.997Z   
2            1             True  2024-03-10T15:15:19.997Z   
3            1             True  2024-03-10T15:15:19.997Z   
4            1             True  2024-03-10T15:15:19.997Z   

   trip_duration_minutes  
0                      9  
1                      9  
2                      9  
3                      9  
4                      9  

[5 rows x 24 columns]

## Outlier

In [ ]:
# import pandas as pd
# from scipy.stats import iqr  # For Interquartile Range (IQR)


# # Calculate IQR (alternative: use standard deviation)
# q1 = df['delay_minutes'].quantile(0.25)
# q3 = df['delay_minutes'].quantile(0.75)
# iqr = q3 - q1

# # Lower and upper bounds for winsorization (replace with appropriate values based on your analysis)
# lower_bound = q1 - 1.5 * iqr
# upper_bound = q3 + 1.5 * iqr

# # Winsorize the 'delay_minutes' column
# df['delay_minutes'] = np.clip(df['delay_minutes'], lower_bound, upper_bound)

import pandas as pd

# Menghitung nilai kuartil pertama (Q1) dan kuartil ketiga (Q3)
Q1 = df['trip_duration_minutes'].quantile(0.25)
Q3 = df['trip_duration_minutes'].quantile(0.75)

# Menghitung rentang interkuartil (IQR)
IQR = Q3 - Q1

# Menentukan batas bawah dan atas untuk menghapus nilai-nilai ekstrem
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Menghapus nilai-nilai ekstrem dari kolom 'delay_minutes'
df = df[(df['trip_duration_minutes'] >= lower_bound) & (df['trip_duration_minutes'] <= upper_bound)]

# Menampilkan DataFrame setelah nilai-nilai ekstrem dihapus
print(df)



      station_id train_id                         line  \
0             AC     2400   COMMUTER LINE TANJUNGPRIUK   
1             AC     2402   COMMUTER LINE TANJUNGPRIUK   
2             AC     2404   COMMUTER LINE TANJUNGPRIUK   
3             AC     2406   COMMUTER LINE TANJUNGPRIUK   
4             AC     2408   COMMUTER LINE TANJUNGPRIUK   
...          ...      ...                          ...   
17751         WT     557A            PRAMBANAN EKSPRES   
17752         YK      551  PRAMBANAN EKSPRES (PRAMEKS)   
17753         YK      553  PRAMBANAN EKSPRES (PRAMEKS)   
17754         YK      555  PRAMBANAN EKSPRES (PRAMEKS)   
17755         YK     557A            PRAMBANAN EKSPRES   

                          route    color destination_x      time_estimated  \
0      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK 2024-05-08 06:07:00   
1      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK 2024-05-08 06:24:00   
2      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK 2024-05-08 06:

In [ ]:
# import pandas as pd
# import matplotlib.pyplot as plt

# # Mengasumsikan data Anda disimpan dalam DataFrame bernama 'df'
# # Kolom yang berisi data keterlambatan dalam menit: 'delay_minutes'

# # Menghitung jumlah kereta untuk setiap durasi keterlambatan
# jumlah_kereta = df['trip_duration'].value_counts()

# # Membuat plot histogram
# plt.hist(df['trip_duration'])

# # Menambahkan label sumbu
# plt.xlabel('Perjalanan (menit)')
# plt.ylabel('Jumlah Kereta')

# # Menambahkan judul
# plt.title('Distribusi Durasi KRL')

# # Menambahkan grid
# plt.grid(axis='y', alpha=0.7)

# # Menampilkan plot
# plt.show()

# # Mencetak informasi statistik dasar
# print('Rata-rata Keterlambatan:', df['trip_duration'].mean(), 'menit')
# print('Median Keterlambatan:', df['trip_duration'].median(), 'menit')
# print('Standar Deviasi Keterlambatan:', df['trip_duration'].std(), 'menit')



In [ ]:
time

station_id train_id                        line  \
0             AC     2400  COMMUTER LINE TANJUNGPRIUK   
1             AC     2402  COMMUTER LINE TANJUNGPRIUK   
2             AC     2404  COMMUTER LINE TANJUNGPRIUK   
3             AC     2406  COMMUTER LINE TANJUNGPRIUK   
4             AC     2408  COMMUTER LINE TANJUNGPRIUK   
...          ...      ...                         ...   
17479         RK      306          COMMUTERLINE MERAK   
17480         RK      308          COMMUTERLINE MERAK   
17481         RK      310          COMMUTERLINE MERAK   
17482         RK      312          COMMUTERLINE MERAK   
17483         RK  D1/314A          COMMUTERLINE MERAK   

                          route    color   destination time_estimated  \
0      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK        6:07:00   
1      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK        6:24:00   
2      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK        6:43:00   
3      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK        7:01:00   
4      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK        7:19:00   
...                         ...      ...           ...            ...   
17479       RANGKASBITUNG-MERAK  #16812B         MERAK       18:55:00   
17480       RANGKASBITUNG-MERAK  #16812B         MERAK        5:05:00   
17481       RANGKASBITUNG-MERAK  #16812B         MERAK        9:55:00   
17482       RANGKASBITUNG-MERAK  #16812B         MERAK       16:45:00   
17483       RANGKASBITUNG-MERAK  #16812B         MERAK       21:22:00   

      destination_time  waktu              updated_at_x  ... AwalLatitude  \
0              6:16:00   Pagi  2024-03-16T17:00:08.063Z  ...     -612.803   
1              6:33:00   Pagi  2024-03-16T17:00:08.063Z  ...     -612.803   
2              6:52:00   Pagi  2024-03-16T17:00:08.063Z  ...     -612.803   
3              7:10:00   Pagi  2024-03-16T17:00:08.063Z  ...     -612.803   
4              7:28:00   Pagi  2024-03-16T17:00:08.063Z  ...     -612.803   
...                ...    ...                       ...  ...          ...   
17479         20:45:00   Sore  2024-03-16T17:00:17.586Z  ...    -6,416667   
17480          6:55:00   Pagi  2024-03-16T17:00:17.586Z  ...    -6,416667   
17481         11:40:00   Pagi  2024-03-16T17:00:17.586Z  ...    -6,416667   
17482         18:30:00   Sore  2024-03-16T17:00:17.586Z  ...    -6,416667   
17483         23:10:00  Malam  2024-03-16T17:00:17.586Z  ...    -6,416667   

           AwalLongi daop_x  fg_enable_x  have_schedule_x  \
0      1.068.032.377      1            1             True   
1      1.068.032.377      1            1             True   
2      1.068.032.377      1            1             True   
3      1.068.032.377      1            1             True   
4      1.068.032.377      1            1             True   
...              ...    ...          ...              ...   
17479     106,133333      1            1             True   
17480     106,133333      1            1             True   
17481     106,133333      1            1             True   
17482     106,133333      1            1             True   
17483     106,133333      1            1             True   

                   updated_at_y AkhirLatitude AkhirLongitude   id  \
0      2024-03-10T15:15:19.997Z   -61.280.924  1.068.270.132  TPK   
1      2024-03-10T15:15:19.997Z   -61.280.924  1.068.270.132  TPK   
2      2024-03-10T15:15:19.997Z   -61.280.924  1.068.270.132  TPK   
3      2024-03-10T15:15:19.997Z   -61.280.924  1.068.270.132  TPK   
4      2024-03-10T15:15:19.997Z   -61.280.924  1.068.270.132  TPK   
...                         ...           ...            ...  ...   
17479  2024-03-10T15:15:19.997Z    -6,1257714     106,656722  MRK   
17480  2024-03-10T15:15:19.997Z    -6,1257714     106,656722  MRK   
17481  2024-03-10T15:15:19.997Z    -6,1257714     106,656722  MRK   
17482  2024-03-10T15:15:19.997Z    -6,1257714     106,656722  MRK   
17483  2024-03-10T15:15:19.

## Waktu Operasi | Pagi/Sore/ Malam

In [ ]:
import pandas as pd

# Konversi kolom waktu menjadi tipe datetime
time['time_estimated'] = pd.to_datetime(time['time_estimated'])
time['destination_time'] = pd.to_datetime(time['destination_time'])

# Mendefinisikan kategori waktu: pagi dan sore hari
dini_hari = time['time_estimated'].dt.hour.between(1,4)
pagi = time['time_estimated'].dt.hour.between(5, 10)
siang = time['time_estimated'].dt.hour.between(11, 14)
sore = time['time_estimated'].dt.hour.between(15, 18)
malam = time['time_estimated'].dt.hour.between(19, 00)

# Menambahkan kolom kategori waktu ke DataFrame
time.loc[dini_hari, 'waktu'] = 'Dini Hari'
time.loc[pagi, 'waktu'] = 'Pagi'
time.loc[siang, 'waktu'] = 'Siang'
time.loc[sore, 'waktu'] = 'Sore'
time.loc[malam, 'waktu'] = 'Malam'

time
# tidak dibahasa inggriskan karena waktu indonesia yaitu dini hari, pagi, siang, sore, malam berbeda dengan in english yaitu siang dan sore sama saja = afternoon

<ipython-input-5-36b2d33db893>:4: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  time['time_estimated'] = pd.to_datetime(time['time_estimated'])
<ipython-input-5-36b2d33db893>:5: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  time['destination_time'] = pd.to_datetime(time['destination_time'])


station_id train_id                        line  \
0             AC     2400  COMMUTER LINE TANJUNGPRIUK   
1             AC     2402  COMMUTER LINE TANJUNGPRIUK   
2             AC     2404  COMMUTER LINE TANJUNGPRIUK   
3             AC     2406  COMMUTER LINE TANJUNGPRIUK   
4             AC     2408  COMMUTER LINE TANJUNGPRIUK   
...          ...      ...                         ...   
17479         RK      306          COMMUTERLINE MERAK   
17480         RK      308          COMMUTERLINE MERAK   
17481         RK      310          COMMUTERLINE MERAK   
17482         RK      312          COMMUTERLINE MERAK   
17483         RK  D1/314A          COMMUTERLINE MERAK   

                          route    color   destination      time_estimated  \
0      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK 2024-05-09 06:07:00   
1      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK 2024-05-09 06:24:00   
2      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK 2024-05-09 06:43:00   
3      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK 2024-05-09 07:01:00   
4      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK 2024-05-09 07:19:00   
...                         ...      ...           ...                 ...   
17479       RANGKASBITUNG-MERAK  #16812B         MERAK 2024-05-09 18:55:00   
17480       RANGKASBITUNG-MERAK  #16812B         MERAK 2024-05-09 05:05:00   
17481       RANGKASBITUNG-MERAK  #16812B         MERAK 2024-05-09 09:55:00   
17482       RANGKASBITUNG-MERAK  #16812B         MERAK 2024-05-09 16:45:00   
17483       RANGKASBITUNG-MERAK  #16812B         MERAK 2024-05-09 21:22:00   

         destination_time  waktu              updated_at_x  ... AwalLatitude  \
0     2024-05-09 06:16:00   Pagi  2024-03-16T17:00:08.063Z  ...     -612.803   
1     2024-05-09 06:33:00   Pagi  2024-03-16T17:00:08.063Z  ...     -612.803   
2     2024-05-09 06:52:00   Pagi  2024-03-16T17:00:08.063Z  ...     -612.803   
3     2024-05-09 07:10:00   Pagi  2024-03-16T17:00:08.063Z  ...     -612.803   
4     2024-05-09 07:28:00   Pagi  2024-03-16T17:00:08.063Z  ...     -612.803   
...                   ...    ...                       ...  ...          ...   
17479 2024-05-09 20:45:00   Sore  2024-03-16T17:00:17.586Z  ...    -6,416667   
17480 2024-05-09 06:55:00   Pagi  2024-03-16T17:00:17.586Z  ...    -6,416667   
17481 2024-05-09 11:40:00   Pagi  2024-03-16T17:00:17.586Z  ...    -6,416667   
17482 2024-05-09 18:30:00   Sore  2024-03-16T17:00:17.586Z  ...    -6,416667   
17483 2024-05-09 23:10:00  Malam  2024-03-16T17:00:17.586Z  ...    -6,416667   

           AwalLongi daop_x  fg_enable_x  have_schedule_x  \
0      1.068.032.377      1            1             True   
1      1.068.032.377      1            1             True   
2      1.068.032.377      1            1             True   
3      1.068.032.377      1            1             True   
4      1.068.032.377      1            1             True   
...              ...    ...          ...              ...   
17479     106,133333      1            1             True   
17480     106,133333      1            1             True   
17481     106,133333      1            1             True   
17482     106,133333      1            1             True   
17483     106,133333      1            1             True   

                   updated_at_y AkhirLatitude AkhirLongitude   id  \
0      2024-03-10T15:15:19.997Z   -61.280.924  1.068.270.132  TPK   
1      2024-03-10T15:15:19.997Z   -61.280.924  1.068.270.132  TPK   
2      2024-03-10T15:15:19.997Z   -61.280.924  1.068.270.132  TPK   
3      2024-03-10T15:15:19.997Z   -61.280.924  1.068.270.132  TPK   
4      2024-03-10T15:15:19.997Z   -61.280.924  1.068.270.132  TPK   
...                         ...           ...            ...  ...   
17479  2024-03-10T15:15:19.997Z    -6,1257714     106,656722  MRK   
17480  2024-03-10T15:15:19.997Z    -6,1257714     106,656722  MRK   
17481  2024-03-10T15:15:19.997Z    -6,1257714     106,656722  MRK   


In [ ]:
time[['time_estimated', 'waktu']].tail(20)

time_estimated  waktu
17464 2024-05-09 18:44:00   Sore
17465 2024-05-09 07:20:00   Pagi
17466 2024-05-09 11:23:00  Siang
17467 2024-05-09 14:45:00  Siang
17468 2024-05-09 18:34:00   Sore
17469 2024-05-09 07:05:00   Pagi
17470 2024-05-09 11:08:00  Siang
17471 2024-05-09 14:17:00  Siang
17472 2024-05-09 18:19:00   Sore
17473 2024-05-09 06:37:00   Pagi
17474 2024-05-09 10:26:00   Pagi
17475 2024-05-09 13:49:00  Siang
17476 2024-05-09 17:50:00   Sore
17477 2024-05-09 07:35:00   Pagi
17478 2024-05-09 13:50:00  Siang
17479 2024-05-09 18:55:00   Sore
17480 2024-05-09 05:05:00   Pagi
17481 2024-05-09 09:55:00   Pagi
17482 2024-05-09 16:45:00   Sore
17483 2024-05-09 21:22:00  Malam

In [ ]:
time.to_csv("clear_data_krl.csv", index=False)

In [ ]:
import pandas as pd

# Konversi kolom waktu menjadi tipe datetime
df['time_estimated'] = pd.to_datetime(df['time_estimated'])
df['destination_time'] = pd.to_datetime(df['destination_time'])

# Mendefinisikan kategori waktu: pagi dan sore hari
dini_hari = df['time_estimated'].dt.hour.between(1,4)
pagi = df['time_estimated'].dt.hour.between(5, 10)
siang = df['time_estimated'].dt.hour.between(11, 14)
sore = df['time_estimated'].dt.hour.between(15, 18)
malam = df['time_estimated'].dt.hour.between(19, 00)

# Menambahkan kolom kategori waktu ke DataFrame
df.loc[dini_hari, 'waktu'] = 'Dini Hari'
df.loc[pagi, 'waktu'] = 'Pagi'
df.loc[siang, 'waktu'] = 'Siang'
df.loc[sore, 'waktu'] = 'Sore'
df.loc[malam, 'waktu'] = 'Malam'

# Menghitung jumlah transaksi per rute dan kategori waktu
jumlah_transaksi_per_rute_dan_waktu = df.groupby(['route', 'waktu']).size()

# Menampilkan hasil
print("Jumlah transaksi per rute dan kategori waktu:")
print(jumlah_transaksi_per_rute_dan_waktu)


NameError: name 'df' is not defined

In [ ]:
# Mendapatkan informasi tanggal
df['date'] = df['time_estimated'].dt.date

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17477 entries, 0 to 17755
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   station_id             17477 non-null  object        
 1   train_id               17477 non-null  object        
 2   line                   17477 non-null  object        
 3   route                  17477 non-null  object        
 4   color                  17477 non-null  object        
 5   destination_x          17477 non-null  object        
 6   time_estimated         17477 non-null  datetime64[ns]
 7   destination_time       17477 non-null  datetime64[ns]
 8   updated_at_x           17477 non-null  object        
 9   name_x                 17477 non-null  object        
 10  AwalLatitude           17477 non-null  float64       
 11  AwalLongi              17477 non-null  float64       
 12  daop_x                 17477 non-null  int64         
 13  fg_ena

In [ ]:
df

station_id train_id                         line  \
0             AC     2400   COMMUTER LINE TANJUNGPRIUK   
1             AC     2402   COMMUTER LINE TANJUNGPRIUK   
2             AC     2404   COMMUTER LINE TANJUNGPRIUK   
3             AC     2406   COMMUTER LINE TANJUNGPRIUK   
4             AC     2408   COMMUTER LINE TANJUNGPRIUK   
...          ...      ...                          ...   
17751         WT     557A            PRAMBANAN EKSPRES   
17752         YK      551  PRAMBANAN EKSPRES (PRAMEKS)   
17753         YK      553  PRAMBANAN EKSPRES (PRAMEKS)   
17754         YK      555  PRAMBANAN EKSPRES (PRAMEKS)   
17755         YK     557A            PRAMBANAN EKSPRES   

                          route    color destination_x      time_estimated  \
0      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK 2024-05-08 06:07:00   
1      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK 2024-05-08 06:24:00   
2      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK 2024-05-08 06:43:00   
3      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK 2024-05-08 07:01:00   
4      JAKARTAKOTA-TANJUNGPRIUK  #DD0067  TANJUNGPRIUK 2024-05-08 07:19:00   
...                         ...      ...           ...                 ...   
17751       YOGYAKARTA-KUTOARJO  #0084D8      KUTOARJO 2024-05-08 18:19:00   
17752       YOGYAKARTA-KUTOARJO  #0084D8      KUTOARJO 2024-05-08 06:37:00   
17753       YOGYAKARTA-KUTOARJO  #0084D8      KUTOARJO 2024-05-08 10:26:00   
17754       YOGYAKARTA-KUTOARJO  #0084D8      KUTOARJO 2024-05-08 13:49:00   
17755       YOGYAKARTA-KUTOARJO  #0084D8      KUTOARJO 2024-05-08 17:50:00   

         destination_time              updated_at_x      name_x  ...  \
0     2024-05-08 06:16:00  2024-03-16T17:00:08.063Z       ANCOL  ...   
1     2024-05-08 06:33:00  2024-03-16T17:00:08.063Z       ANCOL  ...   
2     2024-05-08 06:52:00  2024-03-16T17:00:08.063Z       ANCOL  ...   
3     2024-05-08 07:10:00  2024-03-16T17:00:08.063Z       ANCOL  ...   
4     2024-05-08 07:28:00  2024-03-16T17:00:08.063Z       ANCOL  ...   
...                   ...                       ...         ...  ...   
17751 2024-05-08 18:58:00  2024-03-16T17:00:21.083Z       WATES  ...   
17752 2024-05-08 07:44:00  2024-03-16T17:00:21.080Z  YOGYAKARTA  ...   
17753 2024-05-08 11:47:00  2024-03-16T17:00:21.080Z  YOGYAKARTA  ...   
17754 2024-05-08 15:10:00  2024-03-16T17:00:21.080Z  YOGYAKARTA  ...   
17755 2024-05-08 18:58:00  2024-03-16T17:00:21.080Z  YOGYAKARTA  ...   

       AkhirLatitude  AkhirLongitude   id  daop_y  fg_enable_y  \
0          -6.128092      106.827013  TPK       1            1   
1          -6.128092      106.827013  TPK       1            1   
2          -6.128092      106.827013  TPK       1            1   
3          -6.128092      106.827013  TPK       1            1   
4          -6.128092      106.827013  TPK       1            1   
...              ...             ...  ...     ...          ...   
17751      -7.782344      110.361298  KTA       6            1   
17752      -7.782344      110.361298  KTA       6            1   
17753      -7.782344      110.361298  KTA       6            1   
17754      -7.782344      110.361298  KTA       6            1   
17755      -7.782344      110.361298  KTA       6            1   

      have_schedule_y                updated_at  trip_duration_minutes  waktu  \
0                True  2024-03-10T15:15:19.997Z                      9   Pagi   
1                True  2024-03-10T15:15:19.997Z                      9   Pagi   
2                True  2024-03-10T15:15:19.997Z                      9   Pagi   
3                True  2024-03-10T15:15:19.997Z                      9   Pagi   
4                True  2024-03-10T15:15:19.997Z                      9   Pagi   
...               ...                       ...                    ...    ...   
17751            True  2024-03-10T15:15:19.997Z                     39  Malam   
17752            True  2024-03-10T15:15:19.997Z                   

In [ ]:
df.to_csv("CleanKRLData.csv", index= False)

# Exploratory Data Analysis




## Berdasarkan analisis `daop`, `stasiun`. `route`

1. Berapa jumlah stasiun masing-masing DAOP?
2. Berapa banyak rute yang dilewati oleh stasiun ini?
3. Berapa frekuensi kereta per hari untuk setiap rute?
4. Berapa banyak stasiun yang aktif memiliki jadwal dibandingkan dengan yang tidak?
5. Bagaimana distribusi warna jalur kereta terhadap rute yang dilayani?
6. Berapa lama rata-rata perjalanan dari stasiun awal ke stasiun tujuan?
7.
8. Berapa banyak stasiun yang dilayani oleh lebih dari satu jalur kereta?
9. Apakah ada rute yang memiliki layanan lebih sering di pagi atau sore hari?
10. Bagaimana pengaruh hari kerja vs akhir pekan terhadap jadwal kereta?
11. Bagaimana keakuratan waktu estimasi kedatangan dibandingkan dengan jadwal?
12. Apakah ada stasiun yang menjadi titik transfer utama antar jalur?


In [ ]:
# Menghitung jumlah kereta masing-masing DAOP
jumlah_stasiun_daop = df.groupby('daop')['id_x'].count()

# Menampilkan hasil
print("Jumlah stasiun masing-masing DAOP:")
print(jumlah_stasiun_daop)

KeyError: 'daop'

Jumlah kereta masing-masing DAOP 1  sebanyak  17404 stasiun dan DAOP 6 sebanyak 380

In [ ]:
# Menghitung jumlah rute yang dilewati oleh setiap stasiun
jumlah_rute_per_stasiun = df.groupby('station_id')['route'].nunique()

# Mengurutkan hasil secara menurun
jumlah_rute_per_stasiun_sorted = jumlah_rute_per_stasiun.sort_values(ascending=False)

# Menampilkan hasil
print("Jumlah rute yang dilewati oleh setiap stasiun (diurutkan menurun):")
print(jumlah_rute_per_stasiun_sorted)


Diatas merupakan data dengan jumlah rute pada stasiun terdapat 5 stasiun dengan rute terbanyak yaitu stasiun MRI sebanyak 27 rute, THB sebanyak 22, SUD sebanyak 19, KRI sebanyak 19, dan KAT sebanyak 19 rute.

In [ ]:
# Konversi kolom waktu menjadi tipe datetime
df['time_estimated'] = pd.to_datetime(df['time_estimated'])
df['destination_time'] = pd.to_datetime(df['destination_time'])

# Mendapatkan informasi tanggal
df['date'] = df['time_estimated'].dt.date

# Menghitung frekuensi kereta per hari untuk setiap rute
frekuensi_kereta_per_rute = df.groupby(['route', 'date'])['train_id'].nunique()

# Mengurutkan hasil secara menurun
frekuensi_kereta_per_rute_sorted = frekuensi_kereta_per_rute.sort_values(ascending=False)

# Menampilkan hasil
print("Frekuensi kereta per hari untuk setiap rute (diurutkan menurun):")
print(frekuensi_kereta_per_rute_sorted)


Tabel diatas menunjukkan Frekuensi terbanyak dan tersedikit kereta per hari untuk setiap rute


In [ ]:
# Menghitung jumlah kereta yang memiliki jadwal (have_schedule = TRUE) dan yang tidak
jumlah_kereta_dengan_jadwal = df['have_schedule'].value_counts()[True]
# Menampilkan hasil
print("Jumlah kereta yang memiliki jadwal:", jumlah_kereta_dengan_jadwal)


In [ ]:
# Menghitung jumlah rute yang dilayani oleh setiap warna jalur
distribusi_warna_rute = df.groupby(['color', 'route']).size().reset_index(name='count')

# Membuat plot
plt.figure(figsize=(12, 6))
for color in df['color'].unique():
    data = distribusi_warna_rute[distribusi_warna_rute['color'] == color]
    plt.plot(data['route'], data['count'], marker='o', label=color)

plt.title('Distribusi Warna Jalur Kereta terhadap Rute yang Dilayani')
plt.xlabel('Rute')
plt.ylabel('Jumlah Jalur Kereta')
plt.xticks(rotation=90)
plt.legend(title='Warna Jalur', loc='upper right')
plt.grid(True)
plt.tight_layout()
plt.show()

In [ ]:
# Konversi kolom waktu menjadi tipe datetime
df['time_estimated'] = pd.to_datetime(df['time_estimated'])
df['destination_time'] = pd.to_datetime(df['destination_time'])

# Menghitung durasi perjalanan dari stasiun awal ke stasiun tujuan
df['trip_duration'] = df['destination_time'] - df['time_estimated']

# Menghitung rata-rata durasi perjalanan dari stasiun awal ke stasiun tujuan
rata_rata_durasi_perjalanan = df['trip_duration'].mean()

# Menampilkan hasil
print("Rata-rata durasi perjalanan dari stasiun awal ke stasiun tujuan:", rata_rata_durasi_perjalanan)


In [ ]:
# Konversi kolom waktu menjadi tipe datetime
df['time_estimated'] = pd.to_datetime(df['time_estimated'])
df['destination_time'] = pd.to_datetime(df['destination_time'])

# Menghitung durasi perjalanan dari stasiun awal ke stasiun tujuan dalam menit
df['trip_duration_minutes'] = (df['destination_time'] - df['time_estimated']).dt.total_seconds() / 60

# Menghitung rata-rata durasi perjalanan dari stasiun awal ke stasiun tujuan dalam menit
rata_rata_durasi_perjalanan = df['trip_duration_minutes'].mean()

# Menampilkan hasil
print("Rata-rata durasi perjalanan dari stasiun awal ke stasiun tujuan:", rata_rata_durasi_perjalanan, "menit")


In [ ]:
import pandas as pd


# Menghitung jumlah jalur kereta yang dilayani oleh setiap stasiun
jumlah_jalur_per_stasiun = df.groupby('station_id')['route'].nunique()

# Menghitung jumlah stasiun yang dilayani oleh lebih dari satu jalur kereta
jumlah_stasiun_lebih_dari_satu_jalur = (jumlah_jalur_per_stasiun > 1).sum()

# Menampilkan hasil
print("Jumlah stasiun yang dilayani oleh lebih dari satu jalur kereta:", jumlah_stasiun_lebih_dari_satu_jalur)


In [ ]:
# Menampilkan stasiun-stasiun yang dilayani oleh lebih dari satu jalur kereta
stasiun_lebih_dari_satu_jalur = jumlah_jalur_per_stasiun[jumlah_jalur_per_stasiun > 1].index.tolist()
print("Stasiun yang dilayani oleh lebih dari satu jalur kereta:", stasiun_lebih_dari_satu_jalur)


In [ ]:
import pandas as pd

# Konversi kolom waktu menjadi tipe datetime
df['time_estimated'] = pd.to_datetime(df['time_estimated'])
df['destination_time'] = pd.to_datetime(df['destination_time'])

# Mendefinisikan kategori waktu: pagi dan sore hari
pagi = df['time_estimated'].dt.hour.between(6, 11)
sore = df['time_estimated'].dt.hour.between(12, 17)

# Menambahkan kolom kategori waktu ke DataFrame
df['waktu'] = 'Pagi'
df.loc[sore, 'waktu'] = 'Sore'

# Menghitung jumlah transaksi per rute dan kategori waktu
jumlah_transaksi_per_rute_dan_waktu = df.groupby(['route', 'waktu']).size()

# Menampilkan hasil
print("Jumlah transaksi per rute dan kategori waktu:")
print(jumlah_transaksi_per_rute_dan_waktu)


In [ ]:
import pandas as pd

# Konversi kolom waktu menjadi tipe datetime
df['time_estimated'] = pd.to_datetime(df['time_estimated'])

# Identifikasi hari kerja dan akhir pekan
df['hari'] = df['time_estimated'].dt.dayofweek  # Mendapatkan indeks hari dalam seminggu (0: Senin, 1: Selasa, ..., 6: Minggu)
df['jenis_hari'] = 'Hari Kerja'
df.loc[df['hari'].isin([5, 6]), 'jenis_hari'] = 'Akhir Pekan'

# Hitung jumlah transaksi per jenis hari
jumlah_transaksi_per_jenis_hari = df.groupby('jenis_hari').size()

# Menampilkan hasil
print("Jumlah transaksi per jenis hari:")
print(jumlah_transaksi_per_jenis_hari)


df.info(
)

In [ ]:
# import pandas as pd
# import matplotlib.pyplot as plt

# # Membaca data dari file CSV# Ganti 'nama_file.csv' dengan nama file Anda

# # Membuat histogram distribusi delay_minutes di setiap stasiun
# plt.figure(figsize=(20, 6))
# df.groupby('station_id')['delay_minutes'].plot(kind='hist', bins=20, alpha=0.7, legend=True)
# plt.title('Distribusi Delay_minutes di Setiap Stasiun')
# plt.xlabel('Delay_minutes')
# plt.ylabel('Frequency')
# plt.grid(True)
# plt.show()


## Berdasarkan kolom: `line`, `color`, `destination`, `trip_duration`

Maka, beberapa pertanyaan analisis yang dapat diajukan adalah:

1. Bagaimana distribusi jumlah kereta commuter line berdasarkan daerah commuter line/wilayah operasi?
2. Bagaimana distribusi warna kereta commuter line?
3. Apa tujuan paling umum dari perjalanan kereta commuter line?
4. Bagaimana distribusi waktu keterlambatan kereta commuter line?
5. Bagaimana hubungan antara warna kereta dan daerah tujuan?
6. Apakah terdapat keterkaitan antara daerah commuter line dengan waktu keterlambatan kereta?
7. Apakah terdapat pola tertentu dalam keterlambatan kereta berdasarkan daerah tujuan?

In [ ]:
# Menghitung jumlah kereta commuter line berdasarkan daerah commuter line
jumlah_kereta_per_daerah = df.groupby('line')['train_id'].nunique().sort_values(ascending=False)

# Membuat bar plot distribusi jumlah kereta commuter line berdasarkan daerah commuter line
plt.figure(figsize=(10, 6))
jumlah_kereta_per_daerah.plot(kind='bar', color='skyblue')
plt.title('Distribusi Jumlah Kereta Commuter Line\nBerdasarkan Daerah Commuter Line')
plt.xlabel('Daerah Commuter Line')
plt.ylabel('Jumlah Kereta')
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.show()



*   Distribusi Jumlah Kereta berdasarkan commuterline didominasi oleh commuter line Bogor, Cikarang, RangkasBitung, Tangerang, dan Tanjung Priuk.
*   5 commuter line dengan jumlah kereta plaing sedikit yaitu Prambanan Ekspres, Prameks, Merak, YK, dan BST


In [ ]:
# Kelompokkan data berdasarkan warna kereta dan hitung jumlah kereta untuk setiap warna
jumlah_kereta_per_warna = df.groupby('color')['train_id'].nunique().sort_values(ascending=False)

# Visualisasikan distribusi jumlah kereta commuter line berdasarkan warna
plt.figure(figsize=(10, 6))
jumlah_kereta_per_warna.plot(kind='bar', color=jumlah_kereta_per_warna.index, alpha=0.7)
plt.title('Distribusi Jumlah Kereta Commuter Line\nBerdasarkan Warna Kereta')
plt.xlabel('Warna Kereta')
plt.ylabel('Jumlah Kereta')
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.show()




*   Warna #0084D8 mendominasi jumlah kereta, diikuti dengan #E30A16, #16812B, #623814, dan #DD0067



In [ ]:
# Melakukan pengelompokkan (agregasi) berdasarkan kombinasi warna kereta dan daerah tujuan
grouped_df = df.groupby(['color', 'destination']).size().reset_index(name='count')

# Visualisasikan hubungan antara warna kereta dan daerah tujuan menggunakan diagram batang
plt.figure(figsize=(12, 6))
plt.bar(grouped_df['color'] + ' - ' + grouped_df['destination'], grouped_df['count'], color='skyblue')
plt.title('Hubungan antara Warna Kereta dan Daerah Tujuan')
plt.xlabel('Warna Kereta - Daerah Tujuan')
plt.ylabel('Jumlah Kereta')
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.show()



*   Kereta dengan daerah tujuan Jakarta Kota didominasi dengan warna #E30A16
*   Kereta dengan daerah tujuan Bogor Kota didominasi dengan warna #E30A16
*   Kereta dengan daerah tujuan KampungBandan didominasi dengan warna #0084D8
*   Kereta dengan daerah tujuan Cikarang didominasi dengan warna #0084D8
*   Kereta dengan daerah tujuan Tanah Abang Kota didominasi dengan warna #16812B







*   Kereta paling banyak dengan durasi waktu pada range waktu 30-35 menit sebanyak lebih dari 1600 kereta
*   Sebanyak lebih dari 1600 kereta memiliki  keterlambatan memilki durasi perjalanana pada range waktu 5- 10 menit.
*  lebih dari 1500 kereta memiliki durasi waktu 10-15 menit dan 20-25 menit.
* sebanyak lebih dari 1400 kereta dengan durasi perjalanana pada range 40-45 menit

In [ ]:
# Kelompokkan data berdasarkan daerah commuter line dan hitung rata-rata waktu keterlambatan
grouped_df = df.groupby('line')['trip_duration'].mean().reset_index()

# Visualisasikan data menggunakan diagram batang
plt.figure(figsize=(12, 6))
plt.bar(grouped_df['line'], grouped_df['trip_duration'], color='skyblue')
plt.title('Rata-Rata durasi perjalanan Kereta per Daerah Commuter Line')
plt.xlabel('Daerah Commuter Line')
plt.ylabel('Rata-Rata Waktu Keterlambatan (menit)')
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.show()

Dari hasil analisis rata-rata keterlambatan per daerah commurer line yang telah disajikan, dapat dilihat bahwa:

* MERAK memiliki rata-rata keterlambatan tertinggi sebesar 106.86 menit. Hal ini menunjukkan bahwa perjalanan menuju stasiun MERAK cenderung mengalami keterlambatan yang signifikan.

* RANGKASBITUNG memiliki rata-rata keterlambatan kedua tertinggi, yaitu 58.83 menit. Meskipun tidak sebesar MERAK, namun keterlambatan menuju stasiun RANGKASBITUNG juga tergolong tinggi.

* NAMBO memiliki rata-rata keterlambatan sebesar 52.79 menit, menempatkannya sebagai stasiun tujuan dengan keterlambatan tertinggi ketiga.

* BOGOR dan TANAHABANG juga memiliki rata-rata keterlambatan yang cukup tinggi, masing-masing sebesar 48.69 menit dan 46.08 menit.



In [ ]:
# Kelompokkan data berdasarkan daerah tujuan dan hitung rata-rata durasi perjalanan
grouped_df = df.groupby('destination')['trip_duration'].mean().reset_index()

# Visualisasikan data menggunakan diagram batang
plt.figure(figsize=(12, 6))
plt.bar(grouped_df['destination'], grouped_df['trip_duration'], color='lightgreen')
plt.title('Rata-Rata durasi perjalanan Kereta per Daerah Tujuan')
plt.xlabel('Daerah Tujuan')
plt.ylabel('Rata-Rata Waktu Keterlambatan (menit)')
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.show()



*   Daerah yang paling tingi dengan  waktu keterlambatan dihitung dari  dari rata-rata perjalanan yaitu  100 menit yaituDaerah Tujuan ke Merak dengan  daerah tujuan Merak



## Berdasarkan kolom: `station_id`, `train_id`, `route`, `destination_time`

Maka, beberapa pertanyaan analisis yang dapat diajukan adalah:

1. Bagaimana distribusi jumlah kereta commuter line berdasarkan stasiun?
2. Apa jenis kereta yang paling sering dioperasikan?
3. Bagaimana distribusi rute kereta commuter line?

In [ ]:
# Kelompokkan data berdasarkan stasiun dan hitung jumlah kereta commuter line di setiap stasiun
station_counts = df['station_id'].value_counts()

# Visualisasikan data menggunakan diagram batang
plt.figure(figsize=(15, 9))
station_counts.plot(kind='bar', color='skyblue')
plt.title('Distribusi Jumlah Kereta Commuter Line per Stasiun')
plt.xlabel('Stasiun')
plt.ylabel('Jumlah Kereta')
plt.xticks(rotation=90, ha='right')
plt.tight_layout()
plt.show()



*   Jumlah kereta terbanyak ada pada stasiun MRI, yaitu lebih dari 500 kereta, diikuti dengan stasiun DU sebanyak 350 kereta, stasiun CTA, stasiun DP, stasiun POC
*   Jumlah kereta paling sedikit ada pada stasiun KTA, WT, WJ, JN, dan NMO. Ini perlu digali lagi apa yang mmebuat hal ini terjadi.



In [ ]:
# Hitung jumlah kemunculan setiap jenis kereta
kereta_terbanyak = df['train_id'].value_counts().idxmax()
kereta_tersedikit = df['train_id'].value_counts().idxmin()

print("Kereta yang paling sering dioperasikan adalah:", kereta_terbanyak)
print("Kereta yang paling jarang dioperasikan adalah:", kereta_tersedikit)


In [ ]:
# Kelompokkan data berdasarkan rute kereta dan hitung jumlah kereta dalam setiap kelompok
kereta_per_rute = df.groupby('route')['train_id'].nunique().reset_index()

# Urutkan data berdasarkan jumlah kereta secara descending
kereta_per_rute = kereta_per_rute.sort_values(by='train_id', ascending=False)

# Visualisasikan distribusi jumlah kereta commuter line berdasarkan rute menggunakan diagram batang
plt.figure(figsize=(12, 6))
plt.bar(kereta_per_rute['route'], kereta_per_rute['train_id'])
plt.title('Distribusi Jumlah Kereta Commuter Line Berdasarkan Rute')
plt.xlabel('Rute Kereta')
plt.ylabel('Jumlah Kereta')
plt.xticks(rotation=90, ha='right')
plt.tight_layout()
plt.show()




*   Dari grafik diatas dapat diketahui bahwa rute Jakartakota-Bogor memiliki jumlah kereta terbanyak mendekati 120 kereta, disusul dengan Bogor-Jakartakota dengan selisih yang tidak banyak dari jumlah rute Jakartakota-Bogor, lalu diikuti dengan rute Duri-Tangerang sebanyak 60 kereta, diikuti dengan rute baliknya yaitu Tangerang-Duri dan Tanah Abang-Pangkas Bitung.

*   Dari grafik diatas dapat diketahui bahwa rute Cakung-Cikarang memiliki kereta paling sedikit diantara rute lain, diikuti dengan rute Manggarai-Tanah Abang, Serpong-Parungpanjang, Cikarang-Bekasi, dan Manggarai- Duri.








*   5 kereta dengan durasi perjalanan plaing lama yaitu dengan ID : 1000B. 1002C, 1007B, 1008B, 1009B



In [ ]:
import pandas as pd

# Hitung durasi perjalanan untuk setiap stasiun tujuan dan ubah ke menit
avg_delay_per_destination = df.groupby('destination')['trip_duration'].mean().reset_index()
avg_delay_per_destination['trip_duration_minutes'] = avg_delay_per_destination['trip_duration'].dt.total_seconds() / 60

# Hitung durasi perjalanan untuk setiap rute dan ubah ke menit
avg_delay_per_route = df.groupby('route')['trip_duration'].mean().reset_index()
avg_delay_per_route['trip_duration_minutes'] = avg_delay_per_route['trip_duration'].dt.total_seconds() / 60

# Tampilkan hasil
print("Rata-rata durasi perjalanan per stasiun tujuan (dalam menit):")
print(avg_delay_per_destination.sort_values(by='trip_duration_minutes', ascending=False))

print("\nRata-rata durasi perjalanan per rute (dalam menit):")
print(avg_delay_per_route.sort_values(by='trip_duration_minutes', ascending=False))


Dari hasil analisis rata-rata keterlambatan per stasiun tujuan yang telah disajikan, dapat dilihat bahwa:

* MERAK memiliki rata-rata keterlambatan tertinggi sebesar 106.86 menit. Hal ini menunjukkan bahwa perjalanan menuju stasiun MERAK cenderung mengalami keterlambatan yang signifikan.

* RANGKASBITUNG memiliki rata-rata keterlambatan kedua tertinggi, yaitu 58.83 menit. Meskipun tidak sebesar MERAK, namun keterlambatan menuju stasiun RANGKASBITUNG juga tergolong tinggi.

* NAMBO memiliki rata-rata keterlambatan sebesar 52.79 menit, menempatkannya sebagai stasiun tujuan dengan keterlambatan tertinggi ketiga.

* BOGOR dan TANAHABANG juga memiliki rata-rata keterlambatan yang cukup tinggi, masing-masing sebesar 48.69 menit dan 46.08 menit.



Dari hasil analisis rata-rata keterlambatan per rute yang telah disajikan, dapat dilihat bahwa:

* RANGKASBITUNG-MERAK merupakan rute dengan rata-rata keterlambatan tertinggi, yaitu 106.86 menit. Hal ini menunjukkan bahwa perjalanan dari RANGKASBITUNG ke MERAK memiliki keterlambatan yang signifikan.

* TANAHABANG-RANGKASBITUNG memiliki rata-rata keterlambatan kedua tertinggi, yaitu 60.36 menit. Meskipun tidak sebesar rute RANGKASBITUNG-MERAK, namun keterlambatan pada rute ini juga tergolong tinggi.

* JAKARTAKOTA-NAMBO memiliki rata-rata keterlambatan sebesar 53.69 menit, menempatkannya sebagai rute dengan keterlambatan tertinggi ketiga.

* RANGKASBITUNG-TANAHABANG dan JAKARTAKOTA-BOGOR juga memiliki rata-rata keterlambatan yang cukup tinggi, masing-masing sebesar 51.02 menit dan 50.90 menit.

In [ ]:
import pandas as pd

# Hitung rata-rata keterlambatan untuk setiap stasiun tujuan
avg_delay_per_destination = df.groupby('destination')['trip_duration'].mean().reset_index()

# Hitung rata-rata keterlambatan untuk setiap rute
avg_delay_per_route = df.groupby('route')['trip_duration'].mean().reset_index()

# Tampilkan hasil
print("Rata-rata keterlambatan per stasiun tujuan:")
print(avg_delay_per_destination.sort_values(by='trip_duration', ascending=True))

print("\nRata-rata keterlambatan per rute:")
print(avg_delay_per_route.sort_values(by='trip_duration', ascending=True))


Dari hasil analisis rata-rata keterlambatan per stasiun tujuan yang telah disajikan, dapat dilihat bahwa:

* TANAH ABANG memiliki rata-rata keterlambatan yang paling rendah, yaitu 5.5 menit. Stasiun ini menunjukkan keterlambatan paling minimal dibandingkan stasiun tujuan lainnya.

* BOJONGGEDE memiliki rata-rata keterlambatan kedua terendah, yaitu 9.5 menit. Meskipun sedikit lebih tinggi dari TANAH ABANG, keterlambatan di stasiun ini juga tergolong rendah.

* TANJUNGPRIUK memiliki rata-rata keterlambatan sebesar 12.64 menit, menempatkannya sebagai stasiun tujuan dengan keterlambatan tertinggi ketiga.

* DURI dan TANGERANG juga memiliki rata-rata keterlambatan yang cukup rendah, masing-masing sebesar 15.02 menit dan 16.04 menit.

Dari hasil tersebut, dapat disimpulkan bahwa stasiun tujuan dengan keterlambatan paling rendah adalah TANAH ABANG, sedangkan TANGERANG memiliki keterlambatan paling tinggi di antara stasiun tujuan yang disajikan.

Dari hasil analisis rata-rata keterlambatan per rute yang telah disajikan, dapat dilihat bahwa:

* MANGGARAI-TANAH ABANG merupakan rute dengan rata-rata keterlambatan terendah, yaitu 5.5 menit. Hal ini menunjukkan bahwa perjalanan dari MANGGARAI ke TANAH ABANG memiliki keterlambatan yang minimal.

* TANAHABANG-MANGGARAI memiliki rata-rata keterlambatan kedua terendah, yaitu 8.25 menit.

* TANJUNGPRIUK-JAKARTAKOTA memiliki rata-rata keterlambatan sebesar 9.33 menit, menempatkannya sebagai rute dengan keterlambatan tertinggi ketiga.

* DEPOK-BOJONGGEDE dan DURI-MANGGARAI juga memiliki rata-rata keterlambatan yang cukup rendah, masing-masing sebesar 9.5 menit dan 10.21 menit.

Dari hasil tersebut, dapat disimpulkan bahwa rute dengan keterlambatan paling rendah adalah MANGGARAI-TANAH ABANG, sedangkan TANJUNGPRIUK-JAKARTAKOTA memiliki keterlambatan paling tinggi di antara rute yang disajikan.







# Apri Improvement


### 1. Route dengan kereta paling banyak

In [ ]:
df.groupby('route')['train_id'].nunique().sort_values(ascending=False)

In [ ]:
df.groupby('route')['train_id'].nunique().sort_values(ascending=False).plot(kind='hist')

### 2. Banyaknya stasiun yang dilewati suatu rute perjalanan (banyaknya stasiun yang tergabung dalam suatu rute perjalanan)

In [ ]:
df.info()

In [ ]:
df.groupby('route')['name'].nunique().sort_values(ascending=False)

In [ ]:
df[df['route'] == 'CIKARANG-KAMPUNGBANDAN']['name'].unique()

In [ ]:
df.groupby('route')['name'].nunique().sort_values(ascending=False).plot(kind='hist')
plt.title('Number of Station in Each Route')

### 3. Stasiun dengan jadwal kereta tersibuk

In [ ]:
df.groupby('name')['train_id'].nunique().sort_values(ascending=False)

In [ ]:
df.groupby('name')['train_id'].nunique().sort_values(ascending=False).plot(kind='hist')
plt.title('Number of Train in Each Station')

### 4. Stasiun yang paling sering dilewati suatu rute

In [ ]:
df.groupby('name')['route'].nunique().sort_values(ascending=False)
#Stasiun manggarai setidaknya tergabung dalam 27 rute perjalanan kereta yang berbeda


In [ ]:
df.groupby('name')['route'].nunique().sort_values(ascending=False).plot(kind='hist')
plt.title('Distribution of Number of Route in Each Station')

In [ ]:
# explore manggarai
df[df['name'] == 'MANGGARAI']['route'].value_counts()

### 5. Distribusi Commuter line yang melewati Stasiun & Stasiun dengan commuter line

In [ ]:
df.groupby('name')['line'].nunique().sort_values(ascending=False)

In [ ]:
df.groupby('name')['line'].nunique().sort_values(ascending=False).plot(kind='hist')
plt.title('Distribution of Number of Commuter Line in Each Station')

### 6. Stasiun dengan daftar operasional

In [ ]:
station_df['daop'].value_counts().plot(kind='bar')
plt.title('Daop Distribution')

In [ ]:
station_df[station_df['have_schedule'] == False]['daop'].value_counts().plot(kind='bar')
plt.title('Daop Distribution of Station without Schedule (unactive)')

In [ ]:
station_df[['have_schedule','daop']].value_counts().plot(kind='bar')
plt.title('Daop Distribution of Station by Schedule Availability')

### 7. Stasiun yang masih beroperasi/memiliki jadwal

In [ ]:
station_df['have_schedule'].value_counts().plot(kind='bar')
plt.title('Number of Station by Schedule Availability')

In [ ]:
station_df[station_df['have_schedule'] == False]

### 8. Distribusi jadwal kereta/keberangkatan tiap jam

In [ ]:
# per kereta
df.groupby('time_estimated')['train_id'].nunique().plot(kind='bar')
plt.title('Number of Operating Train by Hour')

In [ ]:
# per jadwal
df.groupby('time_estimated')['id_x'].nunique().plot(kind='bar')
plt.title('Number of Schedule by Hour')

### 9. Korelasi Kesibukan stasiun dengan daftar operasional

In [ ]:
temp = df.copy()
temp = temp.groupby('name')['id_x'].nunique().reset_index(name='no_schedule')
temp1 = pd.merge(temp, station_df, left_on="name", right_on="name", how="inner")
temp1.drop(['id','name','updated_at'], axis=1, inplace=True)
temp1.head()

In [ ]:
temp1.corr(numeric_only=True)

In [ ]:
sns.boxplot(x='daop', y='no_schedule', data=temp1)
plt.title('Number of Schedule by Daop')

In [ ]:
temp1.groupby('daop')['no_schedule'].describe()

### 10. Line terpendek & Line terpanjang (berdasarkan banyaknya stasiun yang dilewati)

In [ ]:
# train_id.groupby('line')['no_station'].describe().sort_values(by='mean', ascending=False)

### 11. Kereta dengan durasi total perjalanan terlama

In [ ]:
# train

In [ ]:
# train['estimated_trip_time'].plot(kind='hist')
# plt.title('Distribution of Estimated Trip Time of each Train')

### 12. Korelasi panjang rute (banyaknya stasiun yang dilewati) dengan waktu tempuh

In [ ]:
# train.select_dtypes("number").corr()

In [ ]:
# sns.scatterplot(x='no_station', y='estimated_trip_time', data=train)

### 13. Distribusi panjang rute

In [ ]:
# train_id.groupby('route')['no_station'].describe().sort_values(by='mean', ascending=False)

In [ ]:
# train.groupby('route')['no_station'].mean().plot(kind='hist')
# plt.title('Distribution of Average Number of Station in Each Route') # Distribusi rata rata angka stasiun per rute = jadi per rute itu ada berapa stasiun gitu

In [ ]:
df.to_csv("krlforviz.csv", index=False)

# Insight

1.
2.
3.
4.
5.
6.
7.
8.
9.
10.
11.
12.
13.
14.
15.
16.
17.
18.
19.
20.